### PR TITLE
Apply path remapping to DW_AT_GNU_dwo_name when producing split DWARF

### PR DESCRIPTION
**Pull request by cbeuw**
`--remap-path-prefix` doesn't apply to paths to `.o` (in case of packed) or `.dwo` (in case of unpacked) files in `DW_AT_GNU_dwo_name`. GCC also has this bug https://gcc.gnu.org/bugzilla/show_bug.cgi?id=91888

### CODE REVIEW

**Comment by oli-obk** [comment]
cc @davidtwco I saw you mention dwp on zulip recently

**Comment by davidtwco** [comment]
I don't have time to do a full review right now (will take a closer look tomorrow), but we could probably do this without breaking packaging:

As rustc invokes the dwarf packaging utility (rather than that being part of a separate build process), we can just always provide the paths to the object files directly rather than rely on the path computed in the packaging utility from `DW_AT_comp_dir` and `DW_AT_GNU_dwo_name` (and for dependencies, we look inside the rlib).

**Review by davidtwco — APPROVED** [review]
Implementation looks good to me, are there other changes you want to make (the PR is marked WIP)?

**Comment by cbeuw** [comment]
I marked it as draft because I didn't really know if this will work and I wanted to see the Linux CI run results (I'm on Windows so can't really test DWARF properly).

I'll remove WIP and I'm happy to merge as-is once bors is happy too

**Comment by cbeuw** [comment]
Turns out my test actually wasn't doing anything and `DW_AT_GNU_dwo_name` still isn't properly remapped.

This acutally came up before #82074. Will look into it later

**Comment by cbeuw** [comment]
The root cause is actually a LLVM bug
https://github.com/rust-lang/rust/blob/0fb1c371d4a14f9ce7a721d8aea683a6e6774f6c/compiler/rustc_codegen_llvm/src/debuginfo/metadata.rs#L1108-L1112

The fix is easy enough. Instead of using the real path to the DWARF file here we use the filename passed in from above in the corresponding `DICompileUnit` instead
https://github.com/llvm/llvm-project/blob/1262cf7f045fd8255f433a3335265fe3bb65c9c4/llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp#L1250-L1253
```
diff --git a/llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp b/llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp
index 52591a18791f..a6478a915f62 100644
--- a/llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp
+++ b/llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp
@@ -1307,9 +1307,9 @@ void DwarfDebug::finalizeModuleInfo() {
                                          : dwarf::DW_AT_GNU_dwo_name;
       finishUnitAttributes(TheCU.getCUNode(), TheCU);
       TheCU.addString(TheCU.getUnitDie(), attrDWOName,
-                      Asm->TM.Options.MCOptions.SplitDwarfFile);
+                      TheCU.getCUNode()->getSplitDebugFilename());
       SkCU->addString(SkCU->getUnitDie(), attrDWOName,
-                      Asm->TM.Options.MCOptions.SplitDwarfFile);
+                      TheCU.getCUNode()->getSplitDebugFilename());
       // Emit a unique identifier for this CU.
       uint64_t ID =
           DIEHash(Asm, &TheCU).computeCUSignature(DWOName, TheCU.getUnitDie());
```

Will submit this to upstream first

cc @nagisa

**Comment by cbeuw** [comment]
LLVM patch: https://reviews.llvm.org/D115485

**Comment by cbeuw** [comment]
I misunderstood how `MCOptions.SplitDwarfFile` is used in LLVM. It does control what will be emitted into `DW_AT_[GNU_]dwo_name`, but it can be remapped because it's not used as a real path if pass in a separate a filestream when we create an emit file pass, which we do here:
https://github.com/rust-lang/rust/blob/4a66a704b2c3d30ff07d89380ebb9ba3de3b3182/compiler/rustc_codegen_llvm/src/back/write.rs#L913-L924

And eventually
https://github.com/rust-lang/rust/blob/4a66a704b2c3d30ff07d89380ebb9ba3de3b3182/compiler/rustc_llvm/llvm-wrapper/PassWrapper.cpp#L644-L656.

So the LLVM change is not needed and the current solution works

File: compiler/rustc_codegen_llvm/src/back/write.rs [hunk 1]
```diff
@@ -901,17 +904,14 @@ pub(crate) unsafe fn codegen(
                     .generic_activity_with_arg("LLVM_module_codegen_emit_obj", &module.name[..]);
 
                 let dwo_out = cgcx.output_filenames.temp_path_dwo(module_name);
-                let dwo_out = match cgcx.split_debuginfo {
```
Comment (davidtwco) [inline compiler/rustc_codegen_llvm/src/back/write.rs:904]: I'm not sure I understand why this change is necessary - the current implementation _is_ a little confused, packed/unpacked should be distinct from split/single, see #89819's re-introduction of `-Zsplit-dwarf-kind` - but that seems unrelated to the change you're making in this pull request.

File: compiler/rustc_codegen_llvm/src/back/write.rs [hunk 2]
```diff
@@ -948,7 +948,7 @@ pub(crate) unsafe fn codegen(
 
     Ok(module.into_compiled_module(
         config.emit_obj != EmitObj::None,
-        cgcx.target_can_use_split_dwarf && cgcx.split_debuginfo == SplitDebuginfo::Unpacked,
```
Comment (davidtwco) [inline compiler/rustc_codegen_llvm/src/back/write.rs:951]: See previous comment.

**Review by davidtwco — CHANGES_REQUESTED** [review]
Looks good to me, comment or two about some changes but broadly what I'd expect.

File: compiler/rustc_codegen_llvm/src/back/write.rs (hunk 1 above)
Comment (cbeuw) [reply compiler/rustc_codegen_llvm/src/back/write.rs:904]: Before, we don't produce the `.dwo` files when it's packed. We rely on `llvm-dwp` to read the paths to `.o` files in the linked executable to read the debuginfo to pack into a single `.dwp`.

The comment above `link_dwarf_object` is said
```
/// Invoke `llvm-dwp` (shipped alongside rustc) to link `dwo` files from Split DWARF into a `dwp`
/// file.
```

So I thought `llvm-dwp` needs to get paths to `.dwo` files, which we weren't producing for `packed`. But I think it's actually reading `.o`. I reverted the changes in this file and gave `llvm-dwp` object files instead and it seems to work

Comment (davidtwco) [reply compiler/rustc_codegen_llvm/src/back/write.rs]: Yeah, that should work, after #89819, the packed/unpacked won't change whether split debuginfo is in dwo files or not, that'll be an independent option.

**Review by davidtwco — APPROVED** [review]
